File: idk.py
import os,pickle
from google_auth_oauthlib.flow import InstalledAppFlow
from google.oauth2.credentials import Credentials
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
import webbrowser
from shutil import which

def wbopentermux(url,new=0,autoraise=True):
    os.system(url)
def wbopennewtermux(url): wbopentermux(url)

def create_service(client_secret_file:str, api_name, api_version, scope, prefix=''):
    CLIENT_SECRET_FILE = client_secret_file
    API_SERVICE_NAME = api_name
    API_VERSION = api_version
    SCOPES = [f'https://www.googleapis.com/auth/{s}' for s in scope]
	
    #termux instance
    if which("termux-open-url") is not None:
        webbrowser.open = wbopentermux
        webbrowser.open_new = wbopennewtermux
    cred = None
    working_dir = os.getcwd()
    token_dir = 'token files'
    pickle_file = f'token_{API_SERVICE_NAME}_{API_VERSION}{prefix}.pickle'

	### Check if token dir exists first, if not, create the folder
    if not os.path.exists(os.path.join(working_dir, token_dir)):
        os.mkdir(os.path.join(working_dir, token_dir))

    if os.path.exists(os.path.join(working_dir, token_dir, pickle_file)):
        with open(os.path.join(working_dir, token_dir, pickle_file), 'rb') as token:
            cred:Credentials = pickle.load(token)

    if not cred or not cred.valid:
        if cred and cred.expired and cred.refresh_token:
            try: cred.refresh(Request())
            except:
                os.remove(os.path.join(working_dir, token_dir, pickle_file))
                flow = InstalledAppFlow.from_client_secrets_file(CLIENT_SECRET_FILE, SCOPES)
                cred = flow.run_local_server(success_message="done",redirect_uri_trailing_slash=False,termux=True)
        else:
            flow = InstalledAppFlow.from_client_secrets_file(CLIENT_SECRET_FILE, SCOPES)
            cred = flow.run_local_server(success_message="done",redirect_uri_trailing_slash=False,termux=True)

        with open(os.path.join(working_dir, token_dir, pickle_file), 'wb') as token:
            pickle.dump(cred, token)
	
	#return cred.token
    try:
        service = build(API_SERVICE_NAME, API_VERSION, credentials=cred)
        return service #fallback to return the snippetless Resource if the file doesnt exist
    except Exception as e:
        print(e)
        print(f'Failed to create service instance for {API_SERVICE_NAME}')
        os.remove(os.path.join(working_dir, token_dir, pickle_file))
        return None

File: idk.js
const fs = require('fs').promises;
const path = require('path');
const process = require('process');
const {authenticate} = require('@google-cloud/local-auth');
const {google} = require('googleapis');

async function create_service(client_secret_file, api_name, api_version, scope, prefix='') {
  // The file token.json stores the user's access and refresh tokens, and is
  // created automatically when the authorization flow completes for the first
  // time.

  //scopes is editable, but it will be out of scopes after this so idc
  let SCOPES = []
  for (let i of scope) {SCOPES.push("https://www.googleapis.com/auth/"+i)}
  const TOKEN_PATH = path.join(process.cwd(), "token files/token_"+api_name+"_"+api_version+".json");
  const CREDENTIALS_PATH = path.join(process.cwd(), client_secret_file);

  /**
   * Reads previously authorized credentials from the save file.
   *
   * @return {Promise<OAuth2Client|null>}
   */
  async function loadSavedCredentialsIfExist() {
    try {
      const content = await fs.readFile(TOKEN_PATH);
      const credentials = JSON.parse(content);
      return google.auth.fromJSON(credentials);
    } catch (err) {
      return null;
    }
  }

  /**
   * Serializes credentials to a file comptible with GoogleAUth.fromJSON.
   *
   * @param {OAuth2Client} client
   * @return {Promise<void>}
   */
  async function saveCredentials(client) {
    const content = await fs.readFile(CREDENTIALS_PATH);
    const keys = JSON.parse(content);
    const key = keys.installed || keys.web;
    const payload = JSON.stringify({
      type: 'authorized_user',
      client_id: key.client_id,
      client_secret: key.client_secret,
      refresh_token: client.credentials.refresh_token,
    });
    await fs.writeFile(TOKEN_PATH, payload);
  }

  /**
   * Load or request or authorization to call APIs.
   *
   */
  async function authorize() {
    let client = await loadSavedCredentialsIfExist();
    if (client) {
      return client;
    }
    client = await authenticate({
      scopes: SCOPES,
      keyfilePath: CREDENTIALS_PATH,
    });
    if (client.credentials) {
      await saveCredentials(client);
    }
    return client;
  }

  return await authorize().then(client => google[api_name]({version:api_version, auth:client}))
}

module.exports.create_service = create_service